CRM/Member/BAO/Membership.php

```diff
@@ -2492,8 +2492,11 @@ public static function recordMembershipContribution(&$params, $ids = array()) {
     // store contribution id
     $params['contribution_id'] = $contribution->id;
 
-    //insert payment record for this membership
-    if (empty($ids['contribution']) || !empty($params['is_recur'])) {
+    // Create membership payment if it does not already exist
+    $membershipPayment = civicrm_api3('MembershipPayment', 'get', [
+      'contribution_id' => $contribution->id,
+    ]);
+    if (empty($membershipPayment['count'])) {
```

Comment:
I guess a contribution could relate to more than one membership - so this might fail to create some

Reply:
@mattwire @MegaphoneJon I think the IF should go altogether - in fact - perhaps by accident -- in current master $ids['contribution'] is ALWAYS empty & also the membershipPayment.create code handles being called twice for the same pair

Reply:
@eileenmcnaughton When we reach this point we are working with a single specific membership ID, the entire function would be called once for each membership so contributions linked to multiple memberships would be ok.
If we remove the IF statement then we call hooks multiple times for membershippayment for no good reason, as if it already exists nothing has changed.  So I'd prefer that the code checks before calling create in this instance.